src/audio/selector/selector.c

```diff
@@ -152,6 +94,111 @@ static void selector_free(struct comp_dev *dev)
 	rfree(dev);
 }
 
+static int selector_verify_params(struct comp_dev *dev,
+				  struct sof_ipc_stream_params *params)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	struct comp_buffer *buffer;
+	struct comp_buffer *sinkb;
+	uint32_t in_channels;
+	uint32_t out_channels;
+
+	tracev_selector_with_ids(dev, "selector_verify_params()");
+
+	sinkb = list_first_item(&dev->bsink_list, struct comp_buffer,
+				source_list);
+
+	/* check whether params->channels (received from driver) are equal to
+	 * cd->config.in_channels_count (PLAYBACK) or
+	 * cd->config.out_channels_count (CAPTURE) set during creating selector
+	 * component in selector_new() or in selector_ctrl_set_data().
+	 * cd->config.in/out_channels_count = 0 means that it can vary.
+	 */
+	if (dev->direction == SOF_IPC_STREAM_PLAYBACK) {
+		/* fetch sink buffer for playback */
+		buffer = list_first_item(&dev->bsink_list, struct comp_buffer,
+					 source_list);
+		if (cd->config.in_channels_count &&
+		    cd->config.in_channels_count != params->channels) {
+			trace_selector_error_with_ids(dev,
+						      "selector_verify_params() error: src in_channels_count does not match pcm channels"
+						     );
+			return -EINVAL;
+		}
+		in_channels = cd->config.in_channels_count;
+
+		/* if cd->config.out_channels_count are equal to 0
+		 * (it can vary), we set params->channels to sink buffer
+		 * channels, which were previosly set in
+		 * pipeline_comp_hw_params()
+		 */
+		out_channels = cd->config.out_channels_count ?
+			cd->config.out_channels_count : buffer->channels;
+		params->channels = out_channels;
+	} else {
+		/* fetch source buffer for capture */
+		buffer = list_first_item(&dev->bsource_list, struct comp_buffer,
+					 sink_list);
+		if (cd->config.out_channels_count &&
+		    cd->config.out_channels_count != params->channels) {
+			trace_selector_error_with_ids(dev,
+						      "selector_verify_params() error: src in_channels_count does not match pcm channels"
+						     );
+			return -EINVAL;
+		}
+		out_channels = cd->config.out_channels_count;
+
+		/* if cd->config.in_channels_count are equal to 0
+		 * (it can vary), we set params->channels to source buffer
+		 * channels, which were previosly set in
+		 * pipeline_comp_hw_params()
+		 */
+		in_channels = cd->config.in_channels_count ?
+			cd->config.in_channels_count : buffer->channels;
+		params->channels = in_channels;
+	}
+
+	/* Set buffer params */
+	buffer_set_params(buffer, params, BUFFER_UPDATE_FORCE);
+
+	/* set component period frames */
+	component_set_period_frames(dev, sinkb->rate);
+
+	/* verify input channels */
+	if (in_channels != SEL_SOURCE_2CH && in_channels != SEL_SOURCE_4CH) {
```

Comment:
these verify channels blocks would be a lot better as a ```switch``` statement as it's easier to add other channel counts and easier to read.

Reply:
@lgirdwood 
I've added switch statement for in/out_channels verification.